### ISSUE TITLE
Forcing Gerbera shutdown by mysql_store_result() failed

### ISSUE THREAD

**Issue by mlaerd** [gerbera/gerbera#1434]
When I run the latest Gerbera using MySQL and try to add directories, I always meet forcing Gerbera shutdown by following error.

```
2021-05-05 21:57:48.527   info: [main.cc:93] logCopyright(): Gerbera UPnP Server v1.8.0-56-g1f9add9e - http://gerbera.io/
...
2021-05-05 21:58:09.200  debug: [update_manager.cc:197] threadProc(): sending updates...
2021-05-05 21:58:09.203  debug: [mysql_database.cc:291] getError(): mysql_error (0): ""
2021-05-05 21:58:09.204  error: [update_manager.cc:206] threadProc(): Fatal error when sending updates: Mysql: mysql_store_result() failed: mysql_error (0): "
"; query: SELECT `id`,`update_id` FROM `mt_cds_object` WHERE `id` IN (1049,1047,1046,1043,1041,1039,1037,1036,1034,1033,1027,1026,1024,1023,1016,1014,1002,100
0,997,996,994,993,992,985,983,981,979,977,976,974,973,971,969,967,966,964,963,957,956,954,953,946,944,942,940,938,937,935,934,928,927,925,924,921,919,917,915,
913,911,909,907,904,902,898,896,20,18,890,888,887,885,998,3,87,286,84,283,280,277,274,272,270,267,264,260,256,254,252,1044,248,244,905,308,306,901,304,899,302
,300,895,298,296,891,294,292,289)
2021-05-05 21:58:09.204  error: [update_manager.cc:207] threadProc(): Forcing Gerbera shutdown.
2021-05-05 21:58:09.204  debug: [update_manager.cc:222] threadProc(): NOT sending updates (string empty or invalid).
...
```

https://github.com/gerbera/gerbera/blob/1f9add9e6bdeade4276bbe6d2a9b98e8146a6725/src/database/mysql/mysql_database.cc#L315-L327

This is a ["NULL mysql_store_result() Return After mysql_query() Success"](https://dev.mysql.com/doc/c-api/8.0/en/null-mysql-store-result.html) issue.
Because mysql_error is 0 and the query is SELECT, it appears to me that "There was a malloc() failure."
Maybe the results of SELECT queries are not freed by mysql_free_result() when the error occurs.

Could you advise me something?

**Comment by KarlStraussberger** [comment]
To me, the message is quite early in the startup process.

Can you check whether there is message like `[mysql_database.cc:262] init(): end` which shows that the database is connected properly, before the error.

**Comment by mlaerd** [comment]
Before I ran, I initialized MySQL DB like following.
```
drop table grb_config_value    ;
drop table  mt_autoscan         ;
drop table  mt_internal_setting ;
drop table  mt_metadata         ;
drop table  mt_cds_object       ;
```
And ran script `src/database/mysql/mysql.sql`

I checked the log and found the database was initialized.

```
2021-05-05 21:57:48.540  debug: [thread_runner.h:94] wait(): ThreadRunner: Waiting for TimerThread
2021-05-05 21:57:48.540  debug: [mysql_database.cc:153] init(): start
2021-05-05 21:57:48.542  debug: [mysql_database.cc:240] init(): db_version: 10
2021-05-05 21:57:48.542  debug: [mysql_database.cc:262] init(): end
2021-05-05 21:57:48.542  debug: [sql_database.cc:2315] doMetadataMigration(): Checking if metadata migration is required
2021-05-05 21:57:48.542  debug: [sql_database.cc:2323] doMetadataMigration(): mt_cds_object rows having metadata: 0
2021-05-05 21:57:48.543  debug: [sql_database.cc:2330] doMetadataMigration(): mt_metadata rows having metadata: 0
2021-05-05 21:57:48.543   info: [sql_database.cc:2337] doMetadataMigration(): About to migrate metadata from mt_cds_object to mt_metadata
2021-05-05 21:57:48.543   info: [sql_database.cc:2338] doMetadataMigration(): No data will be removed from mt_cds_object
2021-05-05 21:57:48.543   info: [sql_database.cc:2354] doMetadataMigration(): Migrated metadata - object count: 0
2021-05-05 21:57:48.543  debug: [sql_database.cc:1721] getConfigValues(): loading 0 items
2021-05-05 21:57:48.543   info: [config_manager.cc:1481] updateConfigFromDatabase(): Loading 0 configuration items from database
2021-05-05 21:57:48.544  debug: [server.cc:98] run(): Starting...
```

```
$ mysql --version
mysql  Ver 15.1 Distrib 10.2.16-MariaDB, for Linux (x86_64) using readline 5.1
```

**Comment by mlaerd** [comment]
After forcing shutdown, when I run Gerbera again, I found some audio albums are registered to DB by browsing Gerbera web page.

**Comment by KarlStraussberger** [comment]
Really strange. I start asking weird questions: What's your distro? Do you have sufficient RAM? Did you get this error before 1.8.0-56?

**Comment by mlaerd** [comment]
> Really strange. I start asking weird questions:
> What's your distro?

I use Plamo Linux (Japanese local distro) derived from Slackware.

> Do you have sufficient RAM?

Yes, 8GB.

> Did you get this error before 1.8.0-56?

I check many versions of [commits](https://github.com/gerbera/gerbera/commits/master?before=1f9add9e6bdeade4276bbe6d2a9b98e8146a6725+105&branch=master) and find;
- Until [v1.7.0-45-g67467d04](https://github.com/gerbera/gerbera/commit/67467d048c9e82dd3172af612dda4ec941359ce5), it works well.
- [v1.7.0-46-g7d69d74b](https://github.com/gerbera/gerbera/commit/7d69d74bb4168acd5da9c835298761e851c10f1e) and later, I get this error.

It seems that transaction does not work correctly on my environment.

**Comment by mlaerd** [comment]
It's very strange for me, too.
It seems that MySQL transaction does not work correctly in my environment .

In v1.7.0-46-g7d69d74b, Gerbera works if I disable the transaction of `SQLDatabase::addObject()` as following;
MySQL error does not occur.

```
$ git diff
diff --git a/src/database/sql_database.cc b/src/database/sql_database.cc
index 7850e50c..585c0fc7 100644
--- a/src/database/sql_database.cc
+++ b/src/database/sql_database.cc
@@ -370,7 +370,7 @@ void SQLDatabase::addObject(std::shared_ptr<CdsObject> obj, int* changedContaine
         throw_std_runtime_error("Tried to add an object with an object ID set");

     std::vector<std::shared_ptr<SQLDatabase::AddUpdateTable>> tables = _addUpdateObject(obj, Operation::Insert, changedContainer);
-    bool withTrans = tables.size() > 1;
+    bool withTrans = false;
     if (withTrans)
         beginTransaction();
     for (const auto& addUpdateTable : tables) {
```

In v1.8.0-56-g1f9add9e, Gerbera works If I disable all transaction like following;.
MySQL error does not occur.

```
$ git diff
diff --git a/src/database/sql_database.cc b/src/database/sql_database.cc
index e219bd5e..0b584995 100644
--- a/src/database/sql_database.cc
+++ b/src/database/sql_database.cc
@@ -485,7 +485,7 @@ void SQLDatabase::addObject(std::shared_ptr<CdsObject> obj, int* changedContaine
         throw_std_runtime_error("Tried to add an object with an object ID set");

     std::vector<std::shared_ptr<SQLDatabase::AddUpdateTable>> tables = _addUpdateObject(obj, Operation::Insert, changedContainer);
-    bool withTrans = tables.size() > 1;
+    bool withTrans = false;
     if (withTrans)
         beginTransaction();
     for (auto&& addUpdateTable : tables) {
@@ -520,7 +520,7 @@ void SQLDatabase::updateObject(std::shared_ptr<CdsObject> obj, int* changedConta
         data = _addUpdateObject(obj, Operation::Update, changedContainer);
     }

-    bool withTrans = data.size() > 1;
+    bool withTrans = false;
     if (withTrans)
         beginTransaction();
     for (auto&& addUpdateTable : data) {
```

**Comment by KarlStraussberger** [comment]
The issue may be that https://github.com/gerbera/gerbera/blob/1f9add9e6bdeade4276bbe6d2a9b98e8146a6725/src/content/update_manager.cc#L203, i.e. https://github.com/gerbera/gerbera/blob/1f9add9e6bdeade4276bbe6d2a9b98e8146a6725/src/database/sql_database.cc#L1243 does not use transactions. So another select can produce the empty result.

If you want to give it a try, just add `beginTransaction()` and `endTransaction()` in update_manager
```
database->beginTransaction();
updateString = database->incrementUpdateIDs(objectIDHash); 
database->endTransaction();
```

**Comment by KarlStraussberger** [comment]
I wanted to go the easy way... In that case, put the calls in the beginning/end of `incrementUpdateIDs`

**Comment by KarlStraussberger** [comment]
> I wanted to go the easy way... In that case, put the calls in the beginning/end of `incrementUpdateIDs` 

And the `endTransaction` is spelled `commit` :smile:

**Comment by mlaerd** [comment]
At first, I confirmed again that v1.8.0-56-g1f9add9e works correctly if all MySQL transaction is disabled as following;
```
diff --git a/src/database/mysql/mysql_database.cc b/src/database/mysql/mysql_database.cc
index 2b24848b..daaa11ef 100644
--- a/src/database/mysql/mysql_database.cc
+++ b/src/database/mysql/mysql_database.cc
@@ -294,12 +294,12 @@ std::string MySQLDatabase::getError(MYSQL* db)

 void MySQLDatabase::beginTransaction()
 {
-    _exec("START TRANSACTION");
+  //_exec("START TRANSACTION");
 }

 void MySQLDatabase::commit()
 {
-    _exec("COMMIT");
+  //_exec("COMMIT");
 }

 std::shared_ptr<SQLResult> MySQLDatabase::select(const char* query, int length)
``` 

Then I tried to insert beginTransaction() and commit() as following;
```
diff --git a/src/database/sql_database.cc b/src/database/sql_database.cc
index e219bd5e..17b3fcc8 100644
--- a/src/database/sql_database.cc
+++ b/src/database/sql_database.cc
@@ -1246,6 +1246,7 @@ std::string SQLDatabase::incrementUpdateIDs(const std::unique_ptr<std::unordered
         return "";
     std::ostringstream inBuf;

+    beginTransaction();
     bool first = true;
     for (auto&& id : *ids) {
         if (first) {
@@ -1278,6 +1279,7 @@ std::string SQLDatabase::incrementUpdateIDs(const std::unique_ptr<std::unordered
         s << row->col(0) << ',' << row->col(1);
         rows.emplace_back(s.str());
     }
+    commit();
     if (rows.empty())
         return "";
     return join(rows, ",");
```

But mysql_error(0) occurs again.

```
2021-05-07 15:20:33.177  debug: [content_manager.cc:1069] addObject(): After adding: parent ID is 4
2021-05-07 15:20:33.177  debug: [thread_runner.h:134] lockGuard(): ThreadRunner: Guard for UpdateThread -
2021-05-07 15:20:33.177  debug: [update_manager.cc:132] containerChanged(): containerChanged. id: 1, signal: true
2021-05-07 15:20:33.177  debug: [update_manager.cc:149] containerChanged(): signalling...
2021-05-07 15:20:33.177  debug: [thread_runner.h:124] notify(): ThreadRunner: Notifying UpdateThread
2021-05-07 15:20:33.177  debug: [update_manager.cc:197] threadProc(): sending updates...
2021-05-07 15:20:33.177  debug: [mysql_database.cc:291] getError(): mysql_error (0): ""
2021-05-07 15:20:33.177  debug: [mysql_database.cc:291] getError(): mysql_error (2014): "Commands out of sync; you can't run this command now"
```

**Comment by KarlStraussberger** [comment]
From the last error message, we can tell that some other transaction is open - I don't know which.

**Comment by KarlStraussberger** [comment]
I am about to create a PR with some more transaction wrappings and an option to disable transactions.

It's here https://github.com/KarlStraussberger/gerbera/tree/mysql-transations

It would be great if you'd try it out

**Comment by mlaerd** [comment]
I appreciate your quick work.
I tried https://github.com/KarlStraussberger/gerbera/tree/mysql-transations .
Although this error occurs with transaction enabled, I can disable transaction by config.xml and confirmed that it works.

Thanks a lot!!

**Comment by KarlStraussberger** [comment]
There's been a lot more than just one transaction. I changed my config and ran into the same issues.

I updated the branch, so if you can confirm that it fixes the issue on your system as well I'll create the Pull Request for the main repository.

**Comment by mlaerd** [comment]
I met compilation error.
"log_sqlite3" is not found.

```
./gerbera/src/database/sqlite3/sqlite_database.cc: In member function 'virtual std::shared_ptr<SQLResult> Sqlite3Database::select(const char*, int)':
./gerbera/src/database/sqlite3/sqlite_database.cc:319:9: error: 'log_sqlite3' was not declared in this scope; did you mean 'sqlite3'?
  319 |         log_sqlite3("Adding select to Queue: {}", query);
      |         ^~~~~~~~~~~
      |         sqlite3
./gerbera/src/database/sqlite3/sqlite_database.cc: In member function 'virtual int Sqlite3Database::exec(const char*, int, bool)':
./gerbera/src/database/sqlite3/sqlite_database.cc:337:9: error: 'log_sqlite3' was not declared in this scope; did you mean 'sqlite3'?
  337 |         log_sqlite3("Adding query to Queue: {}", query);
      |         ^~~~~~~~~~~
      |         sqlite3
./gerbera/src/database/sqlite3/sqlite_database.cc: In static member function 'static void* Sqlite3Database::staticThreadProc(void*)':
./gerbera/src/database/sqlite3/sqlite_database.cc:354:5: error: 'log_sqlite3' was not declared in this scope; did you mean 'sqlite3'?
  354 |     log_sqlite3("Sqlite3Database::staticThreadProc - running thread");
      |     ^~~~~~~~~~~
      |     sqlite3
```

**Comment by mlaerd** [comment]
I tried your latest commit of https://github.com/KarlStraussberger/gerbera/tree/mysql-transations and confirmed that Gerbera works when MySQL transaction is enabled.

Thanks again!

**Comment by mlaerd** [comment]
For your information, I investigate the return address of `mysql_store_result()` by adding debug log in master branch.
I run the Gerbera to add some files to DB and check the log on MySQL transaction enabled and disabled.

When the transaction is disabled, `mysql_res` is changed from 139960501931088 to 139960770441904, pointer address difference is 268,510,816 and no error.
When the transaction is enabled, `mysql_res` is changed from 139715420378144 to 139716024345680, pointer address difference is 603,967,536 and error occurs.

I think this means that when transaction is enabled, memory leak occurs and `mysql_store_result()` is failed by memory allocation failure.


```
diff --git a/src/database/mysql/mysql_database.cc b/src/database/mysql/mysql_database.cc
index 2b24848b..0b731015 100644
--- a/src/database/mysql/mysql_database.cc
+++ b/src/database/mysql/mysql_database.cc
@@ -36,6 +36,10 @@

 #include <netinet/in.h>

+#include <sys/types.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+
 #include "config/config_manager.h"
 #include "util/tools.h"

@@ -321,6 +325,8 @@ std::shared_ptr<SQLResult> MySQLDatabase::select(const char* query, int length)

     MYSQL_RES* mysql_res;
     mysql_res = mysql_store_result(&db);
+    pid_t tid = syscall(__NR_gettid);
+       log_debug("mysql_store_result {} thread {}", (unsigned long)mysql_res, (unsigned long)tid);
     if (!mysql_res) {
         std::string myError = getError(&db);
         throw DatabaseException(myError, "Mysql: mysql_store_result() failed: " + myError + "; query: " + query);
```

**Comment by KarlStraussberger** [comment]
Thanks for your great support.

Your investigation proves my assumption that another `select` was executed in between and broke the the sequence. That's why I added quite some `beginTransaction` calls and the thread blocker. Even the `myTransaction` flag is not the perfect solution but I did not want to rework all the database calls to avoid transaction deadlocks.

Your confirmation that the issue is gone is most valuable so we can savely integrate the patch.
